File: code/js/src/Domain/wrappers/email/EmailValidationError.ts
export abstract class EmailValidationError {
    private readonly defaultMessage: string;

    protected constructor(defaultMessage: string = 'Email is invalid') {
        this.defaultMessage = defaultMessage;
    }

    static Blank = new (class extends EmailValidationError {
        constructor() {
            super('Email cannot be blank');
        }
    })();

    static InvalidFormat = new (class extends EmailValidationError {
        constructor() {
            super('Email has an invalid format');
        }
    })();

    static InvalidLength = class extends EmailValidationError {
        min: number;
        max: number;

        constructor(min: number, max: number) {
            super(`Email must be between ${min} and ${max} characters`);
            this.min = min;
            this.max = max;
        }
    };

    toErrorMessage(): string {
        return this.defaultMessage;
    }
}

File: code/server/jvm/im-domain/src/main/kotlin/im/domain/wrappers/email/EmailValidationError.kt
package im.domain.wrappers.email

sealed class EmailValidationError(
    private val defaultMessage: String = "Email is invalid",
) {
    data object Blank : EmailValidationError("Email cannot be blank")

    data object InvalidFormat : EmailValidationError("Email has an invalid format")

    data class InvalidLength(
        val min: Int,
        val max: Int,
    ) : EmailValidationError("Email must be between $min and $max characters")

    fun toErrorMessage(): String = defaultMessage
}

fun List<EmailValidationError>.toErrorMessage(): String = joinToString("\n") { it.toErrorMessage() }
